### PR TITLE
Generate VirtualServices based on DependencyProxy

### LINKED ISSUE #278: [FEATURE] - Add an ability to provide a mesh service that acts as a proxy through which dependencies should be accessed
**Is your feature request related to a problem? Please describe.**
Need the ability to automatically forward requests to a mesh dependency through a proxy service. 

**Describe the solution you'd like**
Create a new CRD that contains the configuration of how a destination can be reached through a given proxy. The configuration also provides the ability to generate additional hostnames through which a service can be called.

Example CRD:
```
apiVersion: admiral.io/v1alpha1
kind: DependencyProxy
metadata:
  name: dependency-proxy-example
  namespace: admiral
  annotations:
     admiral.io/env: stage
spec:
  destination:
    identity: greeting
    dns_suffix: "xyz"
    dns_prefix: 
        - "test00.foo"
        - "test01.bar"
  proxy:
    identity: nginx-proxy
```

This generates the following ServiceEntry
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: stage.nginx.global-se
spec:
  addresses:
  - 240.0.10.1
  endpoints:
  - address: nginx-gw-svc.default.svc.cluster.local
    locality: us-west-2
    ports:
      http: 80
  hosts:
  - test00.foo.stage.greeting.xyz
  - test01.bar.stage.greeting.xyz
```

### PR DESCRIPTION
**Pull request by shriramsharma**
fixes #278

### CODE REVIEW

File: admiral/pkg/clusters/registry.go [hunk 1]
```diff
@@ -42,6 +42,19 @@ func InitAdmiral(ctx context.Context, params common.AdmiralParams) (*RemoteRegis
 		return nil, fmt.Errorf("error with dependency controller init: %v", err)
 	}
 
+	dependencyProxyHandler := DependencyProxyHandler{
+		RemoteRegistry: w,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/registry.go:46]: Sorry, commenting on an unrelated change: Can we change `w := NewRemoteRegistry(ctx, params)`, and replace `w` with rr`, as we have used `rr` for RemoteRegistry in multiple places.

File: admiral/pkg/clusters/registry.go [hunk 2]
```diff
@@ -42,6 +42,19 @@ func InitAdmiral(ctx context.Context, params common.AdmiralParams) (*RemoteRegis
...
+		dependencyProxyConverter: dependencyProxyConverter{
+			virtualServiceHostNameGenerator:            &virtualServiceHostNameGenerator{},
+			virtualServiceDestinationHostHostGenerator: &virtualServiceDestinationHostHostGenerator{},
```
Comment (nirvanagit) [inline admiral/pkg/clusters/registry.go:49]: May be just call it `virtualServiceDestinationGenerator` ?

File: install/sample/proxy.yaml [hunk 3]
```diff
@@ -0,0 +1,157 @@
+apiVersion: v1
```
Comment (nirvanagit) [inline install/sample/proxy.yaml:1]: How are these resources used?

File: tests/test7.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash -e
```
Comment (nirvanagit) [inline tests/test7.sh:1]: A description of the test will be useful to have.

File: install/sample/proxy.yaml (hunk 3 above)
Comment (shriramsharma) [reply install/sample/proxy.yaml:1]: @nirvanagit , these resources are used in the integration test

File: admiral/pkg/clusters/dependencyproxy.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+
+	cache.put(dependencyProxyObj.Spec.Destination.Identity, fmt.Sprintf("%s-%s", env, dependencyProxyObj.Spec.Destination.Identity), vs)
+	return nil
+}
+
+// generateVSFromDependencyProxy will generate VirtualServices from the configurations provided in the
+// *v1.DependencyProxy object
+func generateVSFromDependencyProxy(ctx context.Context, dependencyProxyObj *v1.DependencyProxy,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxy.go:77]: This function returns `createVirtualServiceSkeleton`, does that function create a working VS based parameters passed to it, or is it just a skeleton?

File: admiral/pkg/clusters/dependencyproxyConverter.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	"fmt"
+	"strings"
+
+	v1 "github.com/istio-ecosystem/admiral/admiral/pkg/apis/admiral/v1"
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/common"
+)
+
+// DependencyProxyConverter provides functions that help convert
+// the given configuration to the corresponding VirtualService
+type DependencyProxyConverter interface {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxyConverter.go:13]: Generic question, why interface of interface?

File: admiral/pkg/clusters/dependencyproxyConverter.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	err := validate(dependencyProxyObj)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate virtual service destination hostname due to error: %w", err)
+	}
+	proxyEnv := dependencyProxyObj.ObjectMeta.Annotations[common.GetEnvKey()]
+	proxyIdentity := dependencyProxyObj.Spec.Proxy.Identity
+	return strings.ToLower(common.GetCnameVal([]string{proxyEnv, proxyIdentity, common.GetHostnameSuffix()})), nil
+}
+
+// GenerateVirtualServiceHostNames generates all the VirtualService's hostnames using the informartion in the
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxyConverter.go:51]: nitpick: typo: `information`

File: admiral/pkg/clusters/dependencyproxyConverter.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	return vsHostNames, nil
+}
+
+func validate(dependencyProxyObj *v1.DependencyProxy) error {
+
+	if dependencyProxyObj == nil {
+		return fmt.Errorf("dependencyProxyObj is nil")
+	}
+	if dependencyProxyObj.ObjectMeta.Annotations == nil {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxyConverter.go:82]: Will it always have `ObjectMeta` in `dependencyProxyObj.ObjectMeta` ?

File: admiral/pkg/clusters/dependencyproxyConverter_test.go [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+	common.ResetSync()
+	common.InitializeConfig(admiralParams)
+
+	testcases := []struct {
+		name               string
+		dependencyProxyObj *v1.DependencyProxy
+		expectedError      error
+	}{
+		{
+			name:               "Given a validating dependency proxy object, when passed dependency proxy obj is nil, the func should return an error",
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxyConverter_test.go:31]: Didn't get the context of `Given a validating dependency proxy object`. Does this object have anything to do with `validating` ?

File: admiral/pkg/clusters/dependencyproxy_test.go [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+					return
+				}
+			}
+
+			if err == nil {
+				if !reflect.DeepEqual(vs.Spec.Hosts, tc.expectedCachedVSHosts) {
+					t.Errorf("expected %v got %v", vs.Spec.Hosts, tc.expectedCachedVSHosts)
+					return
+				}
+				if vs.Spec.Http[0].Route[0].Destination.Host != tc.expectedCachedVSDestinationHost {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/dependencyproxy_test.go:110]: Should we check if object exists before indexing ?

File: admiral/pkg/clusters/handler.go [hunk 11]
```diff
@@ -483,15 +484,16 @@ func handleVirtualServiceEvent(
...
+				// nolint
 				addUpdateVirtualService(ctx, obj, exist, syncNamespace, rc)
 			}
 		}
 	}
 	return nil
 }
 
-func addUpdateVirtualService(ctx context.Context, obj *v1alpha3.VirtualService, exist *v1alpha3.VirtualService, namespace string, rc *RemoteController) {
+func addUpdateVirtualService(ctx context.Context, obj *v1alpha3.VirtualService, exist *v1alpha3.VirtualService, namespace string, rc *RemoteController) error {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler.go:496]: Can we add a function to evaluate if the update is destructive or not? 

And do we need knowledge from multiple clusters to build this virtual service? Just like admiral requires knowledge from multiple clusters to build service entries, hence we skip destructive updates during start up time.

File: admiral/pkg/clusters/serviceentry.go [hunk 12]
```diff
@@ -266,6 +273,42 @@ func modifyServiceEntryForNewServiceOrPod(
...
+		return fmt.Errorf("failed to generate proxy virtual service for sourceIdentity %s as remoteRegistry.AdmiralCache.DependencyProxyVirtualServiceCache is nil", sourceIdentity)
+	}
+	dependencies := remoteRegistry.AdmiralCache.DependencyLookupCache.Get(sourceIdentity)
+	if dependencies == nil {
+		log.Infof("skipped generating proxy virtual service as there are no dependencies found for sourceIdentity %s", sourceIdentity)
+		return nil
+	}
+	for _, dependency := range dependencies {
+		vs := remoteRegistry.AdmiralCache.DependencyProxyVirtualServiceCache.get(dependency)
+		if vs == nil {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/serviceentry.go:290]: nitpick:
```golang
if vs == nil || len(vs) == 0 {
  continue
}
```

File: admiral/pkg/clusters/serviceentry.go [hunk 13]
```diff
@@ -266,6 +273,42 @@ func modifyServiceEntryForNewServiceOrPod(
...
+		for _, v := range vs {
+			existingVS, err := rc.VirtualServiceController.IstioClient.NetworkingV1alpha3().VirtualServices(common.GetSyncNamespace()).Get(ctx, v.Name, v12.GetOptions{})
+			if err != nil && k8errors.IsNotFound(err) {
+				log.Infof("proxy VirtualService %s not found", v.Name)
+			}
+			err = addUpdateVirtualService(ctx, v, existingVS, common.GetSyncNamespace(), rc)
+			if err != nil {
+				return fmt.Errorf("failed generating proxy VirtualService %s due to error: %w", v.Name, err)
+			}
+			log.Infof("successfully generated proxy VirtualService %s", v.Name)
```
Comment (nirvanagit) [inline admiral/pkg/clusters/serviceentry.go:306]: nitpick suggestion: `addUpdateVirtualService` is adding success/failure logs.

File: admiral/pkg/clusters/serviceentry_test.go [hunk 14]
```diff
@@ -2136,3 +2135,176 @@ func TestIsBlueGreenStrategy(t *testing.T) {
...
+		name                               string
+		dependencyLookupCache              *dependencyLookupCache
+		dependencyProxyVirtualServiceCache *dependencyProxyVirtualServiceCache
+		sourceIdentity                     string
+		remoteController                   *RemoteController
+		expectedError                      error
+		expectedVS                         *v1alpha3.VirtualService
+	}{
+		{
+			name:                  "Given dependency proxy to generate VS, when dependencylookipCache is nil, then the func should return an error",
```
Comment (nirvanagit) [inline admiral/pkg/clusters/serviceentry_test.go:2183]: nitpick, type: `dependencylookipCache` -> `dependencylookupCache`

File: admiral/pkg/clusters/serviceentry_test.go [hunk 15]
```diff
@@ -2136,3 +2135,176 @@ func TestIsBlueGreenStrategy(t *testing.T) {
...
+			dependencyLookupCache: nil,
+			expectedError:         fmt.Errorf("remoteRegistry.AdmiralCache.DependencyLookupCache is nil"),
+		},
+		{
+			name:                  "Given dependency proxy to generate VS, when dependencyProxyVirtualServiceCache is nil, then the func should return an error",
+			dependencyLookupCache: &dependencyLookupCache{},
+			expectedError:         fmt.Errorf("remoteRegistry.AdmiralCache.DependencyProxyVirtualServiceCache is nil"),
+		},
+		{
+			name: "Given dependency proxy to generate VS, when the sourceIdentity is not in dependencylookipCache, then the func should not return an error",
```
Comment (nirvanagit) [inline admiral/pkg/clusters/serviceentry_test.go:2193]: Can we also assert that we are calling the client to create the VS ?

File: admiral/pkg/clusters/types.go [hunk 16]
```diff
@@ -124,6 +135,32 @@ func NewRemoteRegistry(ctx context.Context, params common.AdmiralParams) *Remote
 	}
 }
 
+type dependencyLookupCache struct {
+	sourceDestinations map[string][]string
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:139]: should we call this `destinationBySource`, or `sourceToDestinations`?

File: admiral/pkg/clusters/types.go [hunk 17]
```diff
@@ -124,6 +135,32 @@ func NewRemoteRegistry(ctx context.Context, params common.AdmiralParams) *Remote
...
+	mutex              *sync.Mutex
+}
+
+func (d *dependencyLookupCache) put(dependencyObj *v1.Dependency) {
+	if dependencyObj.Spec.Source == "" {
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:144]: Do CRD and sample CR need update? Don't see `Source` in them.

File: admiral/pkg/clusters/dependencyproxyConverter.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+func (*virtualServiceHostNameGenerator) GenerateVirtualServiceHostNames(dependencyProxyObj *v1.DependencyProxy) ([]string, error) {
+	err := validate(dependencyProxyObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate virtual service hostnames due to error: %w", err)
+	}
+	destinationServiceIdentity := dependencyProxyObj.Spec.Destination.Identity
+	dnsSuffix := dependencyProxyObj.Spec.Destination.DnsSuffix
+	proxyEnv := dependencyProxyObj.ObjectMeta.Annotations[common.GetEnvKey()]
+
+	defaultVSHostName := strings.ToLower(common.GetCnameVal([]string{proxyEnv, destinationServiceIdentity, dnsSuffix}))
```
Comment (shriramsharma) [inline admiral/pkg/clusters/dependencyproxyConverter.go:63]: just put this in a interface

File: admiral/pkg/clusters/handler.go (hunk 11 above)
Comment (shriramsharma) [reply admiral/pkg/clusters/handler.go:496]: Add some logging on diff

File: admiral/pkg/clusters/types.go (hunk 16 above)
Comment (shriramsharma) [reply admiral/pkg/clusters/types.go:139]: `sourceToDestinations`

File: admiral/pkg/clusters/dependencyproxyConverter.go (hunk 8 above)
Comment (shriramsharma) [reply admiral/pkg/clusters/dependencyproxyConverter.go:82]: @nirvanagit , `dependencyProxyObj.ObjectMeta` is not a pointer. I added a test to make sure it doesn't panic

File: install/sample/base/kustomization.yaml [hunk 19]
```diff
@@ -1,8 +1,6 @@
 apiversion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
-namespace: sample
```
Comment (nirvanagit) [inline install/sample/base/kustomization.yaml:4]: not using namespace anymore?

File: tests/test7.sh [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+## This test verifies if the DependencyProxy created for the httpbin service
```
Comment (nirvanagit) [inline tests/test7.sh:3]: May be add the flow of request what we discussed yesterday as well.

File: install/sample/base/kustomization.yaml (hunk 19 above)
Comment (shriramsharma) [reply install/sample/base/kustomization.yaml:4]: adding it back, I'm not sure how this got removed

**Review by nirvanagit — APPROVED** [review]
